### DISCUSSION TITLE
features discussions

### DISCUSSION THREAD

**Discussion by theSafdarAwan** [TheSafdarAwan/find-extender.nvim#6] [General]
@kohane27 hi, I saw you commit [here](https://github.com/kohane27/find-extender.nvim/commit/26e57c4ca39c606f91c9f73ebf282b314162d81f#diff-034526f3e1a31f0befc29d2ff0d3adde9693ebcea72d3872dd799c92e80d3c0cL116). Thanks for taking time and reporting this. 

I am implementing something that will address this.
If you have any other suggestions please let me know.

And your suggestion for using `f/F/t/T`, i will add the movement keys option as well. But, i am implementing the `lh` movement as a mode which will allow you to move through the matches using `lh` keys and using the `count` as well.
Also i will add a feature that will allow you to move to the matches in other lines using `jk` keys which will go the match position if there was a match in the line range you specified.
as you said in this [issue](https://github.com/TheSafdarAwan/find-extender.nvim/issues/5) that you use other plugins like flit to go search for match in another line this feature will enable you to do the same thing but using vim motions which gives more control over refining searches according to my experience. 
If you have any suggestions for this feature please let me know.

**Comment by kohane27** [comment]
Hello @TheSafdarAwan, hope you're doing well. Thank you for willing to continue the discussion here. I really appreciate it. For future reference and other people's follow-up, this discussion is spawned from [#5](https://github.com/TheSafdarAwan/find-extender.nvim/issues/5.).

Below is my explanations for "`fFtT` to repeat searches" and suggestions on "lh movement". Please allow me to explain with https://github.com/kohane27/find-extender.nvim/tree/repeat-keys

## `fFtT` to repeat searches

### proof-of-concept with a git commit

git commit: [26a2d11bec1d40816f2a513ee7876171f516d8a0](https://github.com/kohane27/find-extender.nvim/commit/26a2d11bec1d40816f2a513ee7876171f516d8a0). What this commit does is that it allows the user to do the following:

- repeat forward "find" search with `f`
- repeat backward "find" search with `F`
- repeat forward "till" search with `t`
- repeat backward "till" search with `T`

To move the cursor from left to right, use `f` or `t`.
To move the cursor from right to left, use `F` or `T`.

Repeat "find" search with `f` and `t`:

https://user-images.githubusercontent.com/57322459/234799490-8e594596-f8dd-40db-9854-e936e64f9426.mp4

Repeat "till" search with `t` and `T`:

https://user-images.githubusercontent.com/57322459/234799526-1cfc49ba-03be-42be-a50d-f61ac73845f2.mp4

### Text explanation

Let's say we need to find all available "or" in the current line and move the cursor to 2rd "or":

>   I have an **or**ange and the **or**ange is red and the **or**ange is tasty.
>   ^
> cursor

1. press `for` to start the forward "find" search to highlight all possible jumps:

![2023-04-26-16-58-07](https://user-images.githubusercontent.com/57322459/234798952-66916353-eb47-4b51-a20a-9cebaa3ff120.png)

2. press `f` to go to next match

![2023-04-26-16-58-12](https://user-images.githubusercontent.com/57322459/234799015-290838b3-b0e1-4613-9ebb-2f36036b29a9.png)

3. press `f` to go to next match

![2023-04-26-16-58-14](https://user-images.githubusercontent.com/57322459/234799038-64ff29d2-d768-4fb6-9bb3-de9f1e3aac00.png)

4. press `Enter` to exit search with cursor on 3rd "or"

There's an obvious bug in [26a2d11bec1d40816f2a513ee7876171f516d8a0](https://github.com/kohane27/find-extender.nvim/commit/26a2d11bec1d40816f2a513ee7876171f516d8a0). If the user triggers "find" search, the user should not be able to jump between matches with the keys `t` or `T`. The same goes with "till" search: if the user trigger "till" search, the user should not be able to jump between matches with the keys `f` or `F`.

## Why this form of navigation instead of `lh`?

proof-of-concept with a git commit: [26e57c4ca39c606f91c9f73ebf282b314162d81f](https://github.com/kohane27/find-extender.nvim/commit/26e57c4ca39c606f91c9f73ebf282b314162d81f)

### Consistent with other plugins

This way of navigation is the same as [clever-f.vim#usage](https://github.com/rhysd/clever-f.vim#usage) and [flit.nvim](https://github.com/ggandor/flit.nvim). Also, [sneak.vim](https://github.com/justinmk/vim-sneak) has an option to enable the feature of "press the same key to search again":

> *sneak-clever-s*
> Similar to clever-f[3], immediately after invoking Sneak you can move to the
> next match by hitting `s` (or `S`) again. If you instead move the cursor or do
> something else, then `s` starts a new Sneak. In that case, you can repeat the
> most recent Sneak search by pressing "s<Enter>" (or whatever key, if any, you
> mapped to <Plug>Sneak_;). To enable "clever s": >
>     let g:sneak#s_next = 1

Again, I don't use this style of navigation, but I think being consistent with these plugins help attract more users for `find-extender.nvim`.

### More intuitive

Besides being more consistent with other plugin styles, with the above implementation, we don't need to press `<CR>` to confirm the jump. As long as we press a key other than the repeat jump keys (i.e., `f`,`F`,`t`,`T`), we "automatically" escape the search and go back to the normal cursor, and we can directly operate on the word under the cursor.

https://github.com/kohane27/find-extender.nvim/commit/26e57c4ca39c606f91c9f73ebf282b314162d81f

Let's say we need to find all available "or" in the current line and move the cursor to 2rd "or":

>   I have an **or**ange and the **or**ange is red and the **or**ange is tasty.
>   ^
> cursor

1. press `for` to start the forward "find" search to highlight all possible jumps:

![2023-04-26-16-58-07](https://user-images.githubusercontent.com/57322459/234798952-66916353-eb47-4b51-a20a-9cebaa3ff120.png)

2. press `f` to go to next match

![2023-04-26-16-58-12](https://user-images.githubusercontent.com/57322459/234799015-290838b3-b0e1-4613-9ebb-2f36036b29a9.png)

3. press `f` to go to our target match

![2023-04-26-16-58-14](https://user-images.githubusercontent.com/57322459/234799038-64ff29d2-d768-4fb6-9bb3-de9f1e3aac00.png)

4. press, say, `ciw` to directly operate on the word under the cursor (current implementation requires pressing `Enter` to escape the search first)

## Some implementation considerations

### Fake cursor

Currently, with the original `lh` movement implementation, the real cursor stays in its original position during jumping:

![2023-04-26-17-15-09](https://user-images.githubusercontent.com/57322459/234808052-ad5aee0d-6274-4c19-a29a-7c147cc0969b.png)

Red is the real cursor while blue is the fake cursor.

I believe it's a bug with upstream Neovim (https://github.com/ggandor/flit.nvim/issues/3). There's a temporary solution proposed in https://github.com/ggandor/leap.nvim/issues/70. May be it's useful if you'd like to get rid of the dummy cursor solution?

### Configuration options

`init.lua`:

```lua
local DEFAULT_CONFIG = {
    repeat_keys = "ft"
    }
```

May be the configuration can be a pair of keys like above?

```lua
if key == config.repeat_keys[0] or key == config.repeat_keys[0].upper() then
```

The reason is that, the capitalized key is the backward search version. That is, if `repeat_keys = "ri"`:

- repeat forward "find" search with `r`
- repeat backward "find" search with `R`
- repeat forward "till" search with `i`
- repeat backward "till" search with `I`

The user should not need to include the capitalized keys in the config.

## Conclusion

I think the above "`fFtT` to repeat searches" feature is quite different from the current "lh" movement idea. I hope it doesn't make you feel like my suggestion is to make `find-extender.nvim` the "2-key search" version of [clever-f.vim](https://github.com/rhysd/clever-f.vim). I'm already a very happy user of `find-extender.nvim` with the `leap`-like implementation! Thank you again and have a good day!
